### ISSUE TITLE
Add I2P backend support

### ISSUE THREAD

**Issue by suhr** [irungentoo/toxcore#942]
So you could make tox both secure and anonymous.

**Comment by suhr** [comment]
@GrayHatter Reopen please, this is still an issue.

**Comment by TheNain38** [comment]
@GrayHatter Yes, reopen please it didn't get solved

**Comment by ProMcTagonist** [comment]
Would be very cool to have Tox over I2P.

**Comment by JacobHenner** [comment]
I'd be interested in investigating this some more.

**Comment by aaannndddyyy** [comment]
@JacobHenner Do not bundle i2p, just require it as a dependency. The dht would need tweaking (dests instead of ip:port).
use one of these api's:
https://geti2p.net/en/docs/api/samv3
https://geti2p.net/en/docs/api/bob

I2P is a closed network, like tor hidden services, but you could allow nodes to act as bridges just like tcp node now. Would then need a bridge flag so you can find those in the DHT

Tunnels should be ephemeral.
To fully make use of it, use different tunnels (that means different addresses) for DHT communication, 1:1 messenger and groupchat.

**Comment by JacobHenner** [comment]
So would it make more sense to modify toxcore, or implement this in a specific client? Also, if the user is connecting over I2P, will they also be connecting regularly, or is it expected that they will only use one or the other at a time?

**Comment by aaannndddyyy** [comment]
I don't think you could implement a new way of transport in a client. So it'd need to be in core. 
But changes should remain minimal and as generic as possible, so to allow easy adaption for tor and not to bloat core. 
As I said, the hidden services (i2p and tor) are conceived as separate networks.

So if a regular user (non hideen services) were to find his frined on an address such as jf2g5hus6gp2jfgk7zgc2cxtzeedxbstr3ju374amtoaq6p2ax6a.b32.i2p he would not know how to connect to it.

So basically, two networks. those with i2p otion enabled, send this adress to the router and it connects them to their peer.

But for interconnectivity of regular users and hidden servervice users I suggested the "bridges"

**Comment by aaannndddyyy** [comment]
Right now, users who cannot connect via udp find a tcp relay which relays their traffic.
In a similar way, a non-i2p user could find a relay node that is connected to both i2p and regular tox network and thus allow inter-network communication.
But this would be the icing on the cake, in order to prevent fragmentation.
For basic functionality this is not needed, only for inter-network connectivity.

**Comment by GrayHatter** [comment]
any reason you couldn't just plug tox into an I2P proxy like this https://geti2p.net/en/about/browser-config ?

**Comment by JacobHenner** [comment]
@GrayHatter I2P doesn't do "exits" like Tor does. It supports "outproxies" for HTTP sites, but only when the operators are able to do so (and unlike Tor, these proxies typically block non-get requests). 

You can essentially think of I2P as only implementing Tor's "hidden services".

**Comment by GrayHatter** [comment]
@JacobHenner right, so if I used an I2P to bootstrap to a node, also on/in I2P. Wouldn't then I be able to connect to any other client who was also proxied'd and boot strapped to that node?

**Comment by JacobHenner** [comment]
@GrayHatter I don't believe so. Unlike Tor, these nodes don't have IPv4/IPv6 addresses which they could insert into the DHT.

**Comment by aaannndddyyy** [comment]
This is not viable for tox as it is p2p. 
You could of course, proxy it, but then where do you connect to?
Your router does not know your contacts' destinations. This is why sam or bob should be used. Bob is tcp-only, sam can do tcp and udp. 

Think if i2p like tor hidden services. If my tox runs on a tor hidden service but you have only my tox id not my hs address how would you connect to me ( dht part ). If you used the proxy you would still need to specify the contact's destination. 

GrayHatter <notifications@github.com> hat geschrieben:

any reason you couldn't just plug tox into an I2P proxy like this https://geti2p.net/en/about/browser-config ?

**Comment by aaannndddyyy** [comment]
You must handle the addresses. Sam is the easiest way for this. Unless you want the user to manually set up a server tunnel under which he will be reachable.

GrayHatter <notifications@github.com> hat geschrieben:

@JacobHenner right, so if I used an I2P to bootstrap to a node, also on/in I2P. Wouldn't then I be able to connect to any other client who was also proxied'd and boot strapped to that node?

**Comment by aaannndddyyy** [comment]
The new tor hs will also not fit into ipv6 anymore

Jacob Henner <notifications@github.com> hat geschrieben:

@GrayHatter I don't believe so. Unlike Tor, these nodes don't have IPv4/IPv6 addresses which they could insert into the DHT.

**Comment by fcore117** [comment]
I agree that no extra bloat is needed and casual users do not use this, maybe those who want could make fork of Tox or use some workaround https://www.onioncat.org/about-onioncat/

**Comment by aaannndddyyy** [comment]
It would not have to be big changes. Garlicat is better for vpns. It is npt configuration-free. Imho it would be even more complicated if you consider user experience. 
Since i2p would not be bundled it wouls be only minimal changes.

**Comment by TheNain38** [comment]
Yes, it would support communicating to SAM   using the IP and port you give (in the proxy configuration)

**Comment by aaannndddyyy** [comment]
@JacobHenner https://geti2p.net/en/get-involved/develop/applications

**Comment by JacobHenner** [comment]
@aaannndddyyy I'm on it, I'll give it a good look on Friday.

**Comment by aaannndddyyy** [comment]
@JacobHenner thank you. I hope you can come upo with a simple solution that will be robust and tiny, and will be accepted so that no fork is necesarry.

**Comment by JacobHenner** [comment]
Alright, so I definitely am interested in working on this. I'm going to make a branch for it, and experiment over the next few days. I'm thinking that I2P support should be enabled by a ./configure flag, linking to a C I2P library. Clients will have to add support (i.e. a config option somewhere) if they want to enable I2P use. Does this sound reasonable?

**Comment by aaannndddyyy** [comment]
why not in the settings?
if you need to mod the dht for handling dests then all clients should have that mod. 
if you use sam, then both, standard I2P and its imperfect C++ port can be used.

**Comment by TheNain38** [comment]
It could be a new type of proxy: "I2P_PROXY" and the ip and port would be that SAM

**Comment by str4d** [comment]
Hi, (an) I2P developer here! I'm adding myself to the participant list so you all know I'm here listening, and am happy to answer questions.

@GrayHatter One reason for adding native I2P support rather than using one of the proxy tunnels is then you can use UDP directly over I2P via the SAM API. Also, you get control over your own I2P Identity (Destination). Also, it's far easier for your users to set up :)

@JacobHenner There are two ways you could go about it:

1. Use one of the SAM client libraries; that gets you I2P support with minimal bundled code and it works with all I2P implementations, but it requires users to run their router themselves (not a bad thing, I2P routers do better with longer uptime).
2. Bundle i2pd and run the router yourself, which is better for usability for non-I2P-aware users but runs the risk of running two routers on one machine, and also i2pd is much newer and less mature than Java I2P.
3. You do both, and have a config option "Run router myself" which toggles between the internal router and an externally-configured one.

**Comment by aaannndddyyy** [comment]
From the perspective of a (potential) i2p user you'd ideally opt for number 3, but would start with 1.
But from the perspective of the large number of users that will never use i2p, 2 and 3 might not be desired, but only 1.

In any case, I assume, that doing number one would, IMHO, be the best at this point.

**Comment by str4d** [comment]
@aaannndddyyy Yes that is the case. If you require users to use I2PTunnel manually, then you also require them to copy across the tunnel's Destination (either the full Destination or the B32) to report to others. You also can run into problems if the I2P router changes, because I2PTunnel is the place where the keys are stored, and the new I2P router wouldn't have them without getting the user to transfer across the tunnel keys.

With the SAM API (via [the C library](http://git.repo.i2p.xyz/w/libsam3.git)), Tox can create the Destination itself and obtain the private key (as a Base64 string), which it can then use later to create its tunnels with whatever I2P router it happens to be connected to at the time. All the user has to do is configure Tox to know where the SAM API is, and even this can be abstracted away if the router is on the same machine, because the default local ports (7656 for SAM itself, 7655 for sending UDP) are rarely changed.

**Comment by JacobHenner** [comment]
I am going to go with option one for now. I've made some changes to configure.ac to allow the (optional) inclusion of libsam3 when building toxcore, and it seems to work. I'll be examining how to go about adding I2P peers to the Tox DHT this week.

**Comment by timofonic** [comment]
This is really interesting. 

I think one of main concepts of Tox is security, but of course it should become an usable and flexible tool.

My questions:
- Do Tox, i2p and Tor provide different anonymization mechanisms? 
- What about the advantages and disadvantages between Tor and i2p? What's the protocol overhead if combining Tox with one of them?



_My personal experience with optional dependencies:_

They are usually margined and interest gets lost very easily. This happen in all kind of fields, even outside of technology.

In order to understand this, you should emphasize with the thinking from average joe to advanced user spectrum:

- Configuration and maintenance requirements of software: People usually have very limited time availability and very variable but usually too limited attention span. This isn't easy at all, but in order to overcome this I think KISS principle must be applied in a correctly balanced and holistical way.
* Project priorities and principles should become part of the decision taking process: If security is a top priority, use the "secure by default" principle.
* Software bloating: It may not become so much taken into account these days but this can be an adoption burden by slow computer systems: 

**The network effect and a massive arsenal of killer apps might mitigate this and balance to an enough positive result to a limited extent.

** Please considerate slow machines users, advanced users and certain kind of developers: They might be very picky on this, for practical or idealist reasons.

**Comment by JacobHenner** [comment]
> Do Tox, i2p and Tor provide different anonymization mechanisms?

Yes. Tox is not anonymous between contacts. When communicating with another Tox user, your IP is revealed. As for Tor, Tor uses a network of relays and exit nodes, so your connection to Tox appears to be coming from the exit node. With I2P, there are no "exits". Instead, there are tunnels with destination addresses.

> What about the advantages and disadvantages between Tor and i2p? What's the protocol overhead if combining Tox with one of them?

Tox can already be used with Tor. See [this](https://wiki.tox.chat/users/tox_over_tor_tot) for instructions. To use I2P, with Tox, it's my understanding that the Tox DHT would have to be modified to support storing I2P dest addresses.

**Comment by JacobHenner** [comment]
Related: @str4d It's my current understanding that the best way to proceed would be modifying the Tox DHT so it could store I2P destination addresses generated by using SAMv3, and then using the libsam3 library to pass messages between users. Does that sound correct?

**Comment by str4d** [comment]
@JacobHenner essentially yes. You would need space to store 387+ bytes, or 516+ in B64 as-returned by SAM ([see spec for details and explanation of the "+"](https://geti2p.net/en/docs/spec/common-structures#struct_KeysAndCert)).

If you want to store them in a smaller form, you could use the B32 instead, which "looks" like a hostname (`52characters.b32.i2p`), so may be easier to wedge in. It is essentially the SHA-256 of the Destination. This would mean that you would first need to look up the corresponding Destination via `sam3NameLookup()` before connecting with `sam3StreamConnect()` (adding an extra delay to the connection process), and the lookup will only succeed if the remote peer is online.

So it's a tradeoff between DHT size/convenience (B32s are always 60 characters vs. Destinations potentially longer than ~530 bytes depending on signing or crypto algorithm) vs. implementation ease (no naming lookup required if Destinations are stored) and security (B32 risks a collision, although the risk is currently miniscule).

**Comment by JacobHenner** [comment]
Would it be acceptable to add another flag to the Tox DHT, to represent I2P peers? If so, I'll move forward with that plan.

**Comment by aaannndddyyy** [comment]
Wouldn't the different address scheme be indicator enough that you are dealing with an i2p peer?
A flag, however, could be useful in order to mark those nodes that are reachable via both i2p and public internet, so that those could act as gateways/bridges/relays and allow cross-network communication.

**Comment by JacobHenner** [comment]
@aaannndddyyy It would, but in the DHT packets described [here](https://github.com/Tox-Docs/Text/blob/master/src_text/DHT.txt), it appears as if it'd be proper to assign them their own ID, rather than parse them to determine they're not actually the type of address they claim to be.

**Comment by GrayHatter** [comment]
You'd have to do both. But either way, you'd break the packed node format uses if you don't create an extra type.

**Comment by JacobHenner** [comment]
@GrayHatter Do you think it'd be acceptable to create an extra type?

**Comment by GrayHatter** [comment]
@JacobHenner I don't know enough about the differences between I2P and IPv* so say either way. But if you want true I2P support, you'd have to. Rather; I don't know of a better way.

**Comment by GrayHatter** [comment]
So, I've just read a little more about how I2P works, and unless you're planning to implement a I2P tunnel in toxcore. Using a proxy is still the only reasonable solution.

Unless I'm mistaken. Even with I2P address support, you'd still have to send all the I2P traffic to the tunnel on your localhost/domain, so adding I2P address support without implementing a tunnel in toxcore wouldn't serve much purpose.

**Comment by JacobHenner** [comment]
@GrayHatter The idea is to optionally link toxcore to I2P using libsam3 (see https://github.com/JacobHenner/libsam3/). It'd use SAM to communicate with the I2P daemon, and create tunnels to other I2P peers.

**Comment by GrayHatter** [comment]
@JacobHenner My bad, I seem to have missed some of this conversation. Yeah, I see nothing wrong with adding another type to the DHT with a few caveats. 387 bytes is huge compared to the current size of the DHT now, so that's a concern. But how to you expect clients that are compiled without I2P support to interact with clients that do support I2P?

**Comment by JacobHenner** [comment]
> But how to you expect clients that are compiled without I2P support to interact with clients that do support I2P?

I think the current plan was to start with basic I2P-->I2P functionality, and expand it in the future with bridges or something similar.

**Comment by GrayHatter** [comment]
So that would mean you either have toxcore on I2P, or on IPv4/6, but not both?

**Comment by TheNain38** [comment]
@JacobHenner And people that have I2P and IPv* could act as bridges?

**Comment by str4d** [comment]
@GrayHatter as above, if you want to make the DHT entries smaller you can store the B32 instead, but that [has tradeoffs](https://github.com/irungentoo/toxcore/issues/942#issuecomment-157858924).

I2P <--> I2P: this would function essentially exactly as clearnet Tox currently functions, just with the added I2P layer. If implemented on its own with no bridging, this would essentially result in one DHT per transport. In this situation, there is nothing preventing a user from using I2P <--> I2P and IPv* <--> IPv* independently in the same node.

I2P <--> IPv*: My understanding of Tox is that finding peers happens via the DHT, and communicating with them happens via direct connection (correct me if I'm wrong). Then this has two parts:

- Nodes with both I2P and IPv* could bridge the DHT, enabling I2P-only nodes to successfully look up IPv*-only Tox IDs, and vice versa. How this is done should be discussed, because it depends on whether your DHT implementation assumes that a node can always connect to the peers that are storing the DHT entry, or whether you already have recursive lookup support.
  - Vuze (a torrent client that supports simultaneous seeding on I2P and clearnet) participates separately in the two DHTs, so technically you cannot look up clearnet-only torrents, but by the sheer volume of bridging Vuze peers, you can usually find a peer that is dual-seeding. This kind of operation would probably not be as beneficial for Tox, because I assume the goal is to have one single Tox ID space where each Tox ID may be accessible via one or more transports.
- I2P-only nodes will not be able to directly connect to IPv*-only nodes or vice versa, for obvious reasons. This could be handled in two ways:
  - Notify the IPv* user in UI that they should enable I2P support to talk to this Tox ID (the reverse notification for the I2P user would have to be "either drop your anon requirement, or ask that Tox ID to enable I2P support")
  - Add support for connection-proxying nodes, that will route encrypted communication packets between the two networks. This is no different to e.g. using the default false.i2p outproxy for HTTP(S) traffic (where non-I2P requests to the I2P HTTP proxy tunnel get sent to the outproxy). These would be nodes with both I2P and IPv* that additionally enable connection bridging (probably don't want this enabled by default, although doing so would decrease the average load per bridge).

**Comment by JacobHenner** [comment]
Alright, I'm going to go ahead and start implementing the I2P<-->I2P portion first. If that works out well, I'll explore the potential I2P<-->IPv*  bridging.

**Comment by GrayHatter** [comment]
@JacobHenner My strong suggestions is to add plenty of inline comments, and to try to keep everything in the strict I2P <--> I2P and worry about bridging/relaying  later. Also, don't use B32 for the DHT. A collision is worse than slow DHT.

**Comment by JacobHenner** [comment]
@GrayHatter To store the entire destination address, the DHT would have to support variable-length data, as @str4d explained. Would that be permissible (possible?) in the DHT, or would it make more sense to just go ahead and use the b32s?

**Comment by GrayHatter** [comment]
@JacobHenner the DHT already supports VLPackets. You're going to be editing the Packed Node format. That's used by a few places in toxcore. But specifically in the DHT, it already has two sizes one for IPv6 and a smaller size for v4.

**Comment by JacobHenner** [comment]
@GrayHatter Right, but both of those (v4 and v6) have statically defined sizes, whereas the I2P full destination address is variable, because of potential certificate chaining.

**Comment by GrayHatter** [comment]
@JacobHenner @str4d would there be a possibility of padding the size to make it constant?

**Comment by str4d** [comment]
@GrayHatter Technically no. The size of a Destination is variable because it may happen to contain one or more [Certificates](https://geti2p.net/en/docs/spec/common-structures#type_Certificate). In practice, we only use one kind of Certificate - the KeyCertificate, which enables other crypto and signing types to be specified. It is this variable crypto and signing that will really be the potential sticking point in future for length.

One way you *could* make the Destination size a constant is to specify the kind of Destinations that Tox will support - that is, you specify what crypto Tox will allow, and what cert types. Then you would know how long the longest possible Destination would be, and use that for your DHT packet size. It would mean that if in future it became necessary to use a crypto or signing type that required longer key material, you would need to define a new longer packet type.

WRT padding for storing in the DHT, you could pad the Destination object with as many trailing zeroes or random data as you wanted. The Destination object contains [the information necessary to determine its own length](https://geti2p.net/en/docs/spec/common-structures#struct_KeysAndCert), so you could easily parse the packet to strip off the padding (or just use a padding form that encodes its own length).

**Comment by str4d** [comment]
At the moment, a good recommended length is probably 616 bytes for the B64-encoded string (516 for base DSA length + 100 for any certs and extra key material), which would enable you to use the DSA, ECDSA and EdDSA signing types but not RSA. We don't yet know what crypto types we will implement, and how much space they may take up, but we also don't currently use the crypto key in the Destination, so it may not be necessary to extend it any time soon.

**Comment by aaannndddyyy** [comment]
What are the odds for a collision with b32?
Collisions are bad, but if they are extremely unlikely...

**Comment by str4d** [comment]
The B32 is the full SHA-256 hash of the binary Destination, [represented as a 52-character base 32 string](https://geti2p.net/en/docs/naming#base32) (giving a total B32 length of 60 characters). Thus the vulnerability of B32s depends on the collision resistance of SHA-256. However, as the B32s are used to look up Destinations in the netDb, a random collision is not sufficient - the collision must be caused by a valid Destination.

**Comment by GrayHatter** [comment]
@str4d I don't think any risk of collision is acceptable, at least when it comes to Toxcore. But out of curiosity, say sha256 was was broken. How resistant is netDB to poisoning attacks?

**Comment by aaannndddyyy** [comment]
@Grayhatter, @str4d , I _think_ it would be possible not to use netdb at 
all. I have never tried this, but if you have access to your leasesets, 
you can bundle them in the onion such that the other contatc when it 
gets the onion, immediatelly knows the gateways it can connect to. But, 
of course, this would be much bigger then than only a full dest.

**Comment by str4d** [comment]
@GrayHatter If SHA-256 was broken so badly that a valid Destination could be crafted with a hash matching a target, then the netDb would not be very reliable any more, because the DHT routing key *is* the SHA-256 hash of the Destination. It would allow a persistent attacker to take over a B32 by pushing their own LeaseSets, which would easily affect new clients looking up the B32. It would not be 100% effective though, because LeaseSets must be re-published every 10 minutes, and the newest one wins. (That in fact is how multihoming works in I2P - two routers run the same Destination, and whichever one happens to have published a LeaseSet to a particular floodfill most recently will be chosen by clients querying that floodfill.)

However, I am not sure right now how existing clients would be affected by a SHA-256 break, ie. those that already had the valid Destination, and then re-looked up a LeaseSet for that Destination and received one with a different Destination. I imagine that should throw up all sorts of red flags, but I'd need to go check the source code.

Bottom line is, if your packet can only hold a B32, you directly take on the risk of a full SHA-256 break. If you store the full Destination, you pass that risk back to I2P; it is still there, but you then benefit if and when we alter the netDb under the covers to mitigate the risk in future.

I will say that I2P is certainly better than Tor in this respect, at least at present, because they use a truncated SHA-1 hash for their .onion addresses. HS 2.0 actually proposes moving to a 52-character identifier like I2P, although they are planning some crypto blinding magics in there too (some of which I am investigating for potential use in I2P).

@aaannndddyyy I assume by "onion" you mean the Tox DHT packet? I guess if you wanted to, you could essentially publish the LeaseSets in the Tox DHT, but I'm not sure the extra effort to get around using the netDb would be worth any gains.

**Comment by aaannndddyyy** [comment]
@str4d It would avoid the sha issue altogether. Also the onion routing of the onion packet prevents the dht or netdb from learning who communicates with whom.

Using two DHT's for connecting to a peer when one would suffice sounds a bit suboptimal. They main issue with doing it netdb-less would, imho, be the leaseset sizenull

**Comment by str4d** [comment]
@aaannndddyyy Ah, I understand what you mean now.

The netDb is only used for initial contact, ie. determining the current LeaseSet of the remote peer when your node wants to connect, if you don't have a currently-valid one locally. Once a connection is established, the I2P routers will bundle subsequent new LeaseSets in their communications alongside their data as GarlicCloves.

I2P already does LeaseSet lookups in the netDb through a node's client tunnels, specifically for the reason you state.

**Comment by aaannndddyyy** [comment]
What I'm saying is that tox already does a lppkup in a dht. Sp maybe we wang to avoid doing a second lookup in a second dht. Also tox dht has provisions to make it harder for participants of the dht to see which ip wants to cpnnect to which ip, or here which anon identity with which other anon identity. This second aspect is less important in case you use ephemeral destinations, which I'd recommend anyway.

**Comment by fcore117** [comment]
question, if developers plan to integrate this into official Tox then is i2p disabled by default and even then do not use not one bit of extra network resources. I ask because of two reasons, one is that Tox goal is mostly for casual people in world and second Tox currently already has quite large network usage/activity.

If this i2p addon will be planned as fork then no worry.

**Comment by aaannndddyyy** [comment]
This will not add i2p to toxcore but only support for i2p trasport, using an external i2p router which the user would a) have to install and b) configurw tox to use it, in order to use tox over i2p.
That means that unless you yourself explicitly configure tox to use i2p AND you install i2p, there will npt a single additipnal bit be used by your tox. Except fpr the data stpred in the dht, which, due to longer addresses, wpuld slightly increase. That is similar tp the increase ypu get for allowing IPv6 addresses instead of only IPv4.

I hope that answers your question.

**Comment by aaannndddyyy** [comment]
Ideally it would not remain a fork, because the goal is global connectivity, whether you are on plain IP, tor or i2p.
Right mow toxcore allows proxying through tor already and it does not increase traffic for those who do not use tor, and it is mot a fork that tor users would have to use.
Think pf i2p as an analogous case.

**Comment by aaannndddyyy** [comment]
@JacobHenner Any news?

**Comment by fcore117** [comment]
if i am not mistaken they want to complete critical everyday features for casual people who need Skype replacement and later these extra features come to table for thinking. Developers correct if i am wrong.

**Comment by GrayHatter** [comment]
@str4d @JacobHenner any updates on this?

**Comment by str4d** [comment]
Just FYI, I2P has launched [a development program](https://geti2p.net/en/blog/post/2016/06/01/I2P-Summer-Dev) aimed at helping developers integrate I2P into their apps. If you'd like some help getting this moving, let me know!

**Comment by GrayHatter** [comment]
I'll echo @str4d I've become a LOT more familiar with tox DHT so while I don't have the time to take on this project my self. I'm happy to spend as much time as I can getting someone else up to speed on this.

@JacobHenner you still alive mate?

**Comment by JacobHenner** [comment]
@GrayHatter Still alive, but I do not have the resources to dedicate to this effort. Sorry!
